sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java

```diff
@@ -23,40 +23,51 @@
 package io.crate.expression.scalar;
 
 import io.crate.data.Input;
-import io.crate.metadata.BaseFunctionResolver;
 import io.crate.metadata.FunctionIdent;
-import io.crate.metadata.FunctionImplementation;
 import io.crate.metadata.FunctionInfo;
 import io.crate.metadata.Scalar;
 import io.crate.metadata.TransactionContext;
-import io.crate.metadata.functions.params.FuncParams;
-import io.crate.metadata.functions.params.Param;
+import io.crate.types.CollectionType;
 import io.crate.types.DataType;
 import io.crate.types.DataTypes;
 import io.crate.types.ObjectType;
 import io.crate.types.StringType;
+import io.crate.types.UndefinedType;
 
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
 
 public class SubscriptObjectFunction extends Scalar<Object, Map> {
 
-    public static final String NAME = "subscript_obj";
+    private static final String NAME = "subscript_obj";
     private FunctionInfo info;
 
     public static void register(ScalarFunctionModule module) {
-        module.register(NAME,
-            new BaseFunctionResolver(
-                FuncParams.builder(Param.of(ObjectType.untyped()), Param.of(StringType.INSTANCE)).build()
-            ) {
-                @Override
-                public FunctionImplementation getForTypes(List<DataType> dataTypes) throws IllegalArgumentException {
-                    return new SubscriptObjectFunction(
-                        new FunctionInfo(new FunctionIdent(NAME, dataTypes.subList(0, 2)), DataTypes.UNDEFINED));
-                }
+        for (DataType dataType : DataTypes.allRegisteredTypes()) {
+            module.register(ofReturnType(dataType));
+        }
+    }
+
+    public static SubscriptObjectFunction ofReturnType(DataType returnType) {
+        return new SubscriptObjectFunction(
+            new FunctionInfo(
+                new FunctionIdent(getNameForReturnType(returnType), List.of(ObjectType.untyped(), StringType.INSTANCE)),
+                returnType));
+    }
+
+    public static String getNameForReturnType(DataType dataType) {
```

Comment:
I think this method could use some explanation. Why is this necessary?

Reply:
See my previous comment. Our function registration does not take the function return type into account. I solved this by registering the same function with a different return type under different names. Any other suggestion is appreciated.

Reply:
Sorry I think I wasn't clear - I think this aspect of using the type name within the function should be documented (explaining why it is necessary) in the code somewhere. Either on the class level or here on this method.

Reply:
Ah ok, right will do.

Reply:
I've added class level documentation.